### ISSUE TITLE
Remove booked slots from Select Appointment (page2)

### ISSUE THREAD

**Issue by wvdvegt** [alextselegidis/easyappointments#826]
Hi

Is it possible to hide appointments slots that are already filled with the max number of attendants?

When I try to book an already booked slot again I get a popup at page 3 (so after tediously entering my details and pressing next at page 3. So the workflow is a bit strange.

The dialog is also not very clear how to dismiss (seems it responds to the escape key after some experimenting). A close button would not harm.

**Comment by wvdvegt** [comment]
I think I have spotted an issue which might explain. 

When I book 12:00pm in a Amsterdam+1 TZ the item 2:00pm in the available slots list is missing.  The providers TZ was UTC (so 2 hrs off atm i think). When I matched the TZ of the provider with the one displayed in 'Select Appointment Date And Time' the already occupied slots are correctly removed/hidden from display. 

Seems to be caused in frontend_book_api.js by the lines (I think):

                        var availableHourMoment = moment
                            .tz(selDate + ' ' + availableHour + ':00', providerTimezone)
                            .tz(selectedTimezone);

**Comment by wvdvegt** [comment]
Here a patch to correct TimeZone issues leading to the issues described (wrong appointment slots are removed when TimeZones of client and provider do not match). Tested with a provider in Azores TZ and a customer in Amsterdam TZ.

```
diff --git a/application/controllers/Appointments.php b/application/controllers/Appointments.php
index 09c8b61d..b1d254a8 100755
--- a/application/controllers/Appointments.php
+++ b/application/controllers/Appointments.php
@@ -417,7 +417,7 @@ class Appointments extends CI_Controller {
                 $selected_date, $exclude_appointments);
 
             $available_hours = $this->calculate_available_hours($empty_periods, $selected_date,
-                $service['duration'], $service['availabilities_type']);
+                $service['duration'], $provider, $service['availabilities_type']);
 
             if ($service['attendants_number'] > 1)
             {
@@ -497,7 +497,7 @@ class Appointments extends CI_Controller {
                     $empty_periods = $this->get_provider_available_time_periods($provider['id'], $selected_date);
 
                     $available_hours = $this->calculate_available_hours($empty_periods, $selected_date,
-                        $service['duration'], $service['availabilities_type']);
+                        $service['duration'], $provider, $service['availabilities_type']);
 
                     if ($service['attendants_number'] > 1)
                     {
@@ -539,6 +539,9 @@ class Appointments extends CI_Controller {
         $excluded_appointment_ids = []
     )
     {
+        $provider = $this->providers_model->get_row($provider_id);
+        $provider_timezone = new DateTimeZone($provider['timezone']);
+
         // Get the service, provider's working plan and provider appointments.
         $working_plan = json_decode($this->providers_model->get_setting('working_plan', $provider_id), TRUE);
 
@@ -656,13 +659,16 @@ class Appointments extends CI_Controller {
                 $appointment_start = new DateTime($provider_appointment['start_datetime']);
                 $appointment_end = new DateTime($provider_appointment['end_datetime']);
 
+                $appointment_start->setTimezone($provider_timezone);
+                $appointment_end->setTimezone($provider_timezone);
+
                 if ($appointment_start >= $appointment_end)
                 {
                     continue;
                 }
 
-                $period_start = new DateTime($selected_date . ' ' . $period['start']);
-                $period_end = new DateTime($selected_date . ' ' . $period['end']);
+                $period_start = new DateTime($selected_date . ' ' . $period['start'], $provider_timezone);
+                $period_end = new DateTime($selected_date . ' ' . $period['end'], $provider_timezone);
 
                 if ($appointment_start <= $period_start && $appointment_end <= $period_end && $appointment_end <= $period_start)
                 {
@@ -752,15 +758,18 @@ class Appointments extends CI_Controller {
         array $empty_periods,
         $selected_date,
         $service_duration,
+        $provider,
         $availabilities_type = 'flexible'
     )
     {
         $available_hours = [];
 
+        $provider_timezone = new DateTimeZone($provider['timezone']);
+
         foreach ($empty_periods as $period)
         {
-            $start_hour = new DateTime($selected_date . ' ' . $period['start']);
-            $end_hour = new DateTime($selected_date . ' ' . $period['end']);
+            $start_hour = new DateTime($selected_date . ' ' . $period['start'],$provider_timezone);
+            $end_hour = new DateTime($selected_date . ' ' . $period['end'],$provider_timezone);
             $interval = $availabilities_type === AVAILABILITIES_TYPE_FIXED ? (int)$service_duration : 15;
 
             $current_hour = $start_hour;
@@ -795,6 +804,8 @@ class Appointments extends CI_Controller {
         $provider
     )
     {
+        $provider_timezone = new DateTimeZone($provider['timezone']);
+
         $unavailability_events = $this->appointments_model->get_batch([
             'is_unavailable' => TRUE,
             'DATE(start_datetime)' => $selected_date,
@@ -807,8 +818,8 @@ class Appointments extends CI_Controller {
 
         $periods = [
             [
-                'start' => new DateTime($selected_date . ' ' . $working_hours['start']),
-                'end' => new DateTime($selected_date . ' ' . $working_hours['end'])
+                'start' => new DateTime($selected_date . ' ' . $working_hours['start'],$provider_timezone),
+                'end' => new DateTime($selected_date . ' ' . $working_hours['end'],$provider_timezone)
             ]
         ];
 
@@ -1176,6 +1187,8 @@ class Appointments extends CI_Controller {
         $this->load->model('services_model');
         $this->load->model('appointments_model');
 
+        $this->load->model('providers_model');
+
         $post_data = $this->input->post('post_data');
 
         $appointment = $post_data['appointment'];
@@ -1216,6 +1229,9 @@ class Appointments extends CI_Controller {
             return $appointment['id_users_provider'];
         }
 
+        $provider = $this->providers_model->get_row($appointment['id_users_provider']);
+        $provider_timezone = new DateTimeZone($provider['timezone']);
+
         $available_periods = $this->get_provider_available_time_periods(
             $appointment['id_users_provider'],
             date('Y-m-d', strtotime($appointment['start_datetime'])),
@@ -1226,9 +1242,11 @@ class Appointments extends CI_Controller {
         foreach ($available_periods as $period)
         {
             $appt_start = new DateTime($appointment['start_datetime']);
+            $appt_start->setTimezone($provider_timezone);
             $appt_start = $appt_start->format('H:i');
 
             $appt_end = new DateTime($appointment['start_datetime']);
+            $appt_end->setTimezone($provider_timezone);
             $appt_end->add(new DateInterval('PT' . $service_duration . 'M'));
             $appt_end = $appt_end->format('H:i');
 
@@ -1305,7 +1323,7 @@ class Appointments extends CI_Controller {
                         $current_date->format('Y-m-d'), $exclude_appointments);
 
                     $available_hours = $this->calculate_available_hours($empty_periods, $current_date->format('Y-m-d'),
-                        $service['duration'], $service['availabilities_type']);
+                        $service['duration'], $curr_provider, $service['availabilities_type']);
 
                     if ( ! empty($available_hours))
                     {



```

In the mailing the TZ of the provider is mentioned yet times are the customer ones . Also an old/deprecated timezone_model is loaded, preventing e-mail.

```
diff --git a/engine/Notifications/Email.php b/engine/Notifications/Email.php
index d028c725..3b82fa17 100755
--- a/engine/Notifications/Email.php
+++ b/engine/Notifications/Email.php
@@ -86,10 +86,7 @@ class Email {
         Text $icsStream
     ) {
         $framework = get_instance();
-
-        $framework->load->model('timezones_model');
-
-        $timezones = $framework->timezones_model->to_array();
+        $timezones = $framework->timezones->to_array();
 
         switch ($company['date_format'])
         {
@@ -125,7 +122,7 @@ class Email {
         $appointment_provider = $provider['first_name'] . ' ' . $provider['last_name'];
         $appointment_start_date = date($date_format . ' ' . $timeFormat, strtotime($appointment['start_datetime']));
         $appointment_end_date = date($date_format . ' ' . $timeFormat, strtotime($appointment['end_datetime']));
-        $appointment_timezone = $timezones[$provider['timezone']];
+        $appointment_timezone = $timezones[$customer['timezone']];
         $appointment_link = $appointmentLink->get();
         $company_link = $company['company_link'];
         $company_name = $company['company_name'];
@@ -181,10 +178,7 @@ class Email {
         Text $reason
     ) {
         $framework = get_instance();
-
-        $framework->load->model('timezones_model');
-
-        $timezones = $framework->timezones_model->to_array();
+        $timezones = $framework->timezones->to_array();
 
         switch ($company['date_format'])
         {
```

**Comment by wvdvegt** [comment]
Any update on this issue/bug?

**Comment by wvdvegt** [comment]
See #871 for a more recent patch.

**Comment by wvdvegt** [comment]
Any ETA on applying this bugfix?

**Comment by alextselegidis** [comment]
This one is now fixed for v1.4.0 
<small>
<table>
<tr>
<td><img src="https://easyappointments.org/img/easyappointments-signature.png"></td>
<td>&nbsp;&nbsp;Alex Tselegidis, Easy!Appointments Creator <br>
&nbsp;&nbsp;<a href="https://alextselegidis.com">Need a customization? Contact me in person!</a></td>
</tr>
</table>
</small>
